larq_compute_engine/tflite/kernels/bconv2d_impl.h

```diff
@@ -113,29 +116,45 @@ inline void BConv2D(const ConvParams& params, const RuntimeShape& input_shape,
   int m = 0;
   int k = 0;
 
-  // Buffer for bitpacked input data
-  static std::vector<TBitpacked> input_data_bp;
-
   if (bitpack_before_im2col) {
-    // Input tensor has this shape which we bitpack along the channels dimension
-    //  [batch, input height, input width, channels]
-    RuntimeShape packed_input_shape;
-    packbits_tensor(input_shape, input_data, packed_input_shape, input_data_bp);
-
-    // Filter tensor was already bitpacked. Only get the new shape
+    // The filter tensor was already bitpacked. Only get the new shape.
     RuntimeShape packed_filter_shape = packed_shape<TBitpacked>(filter_shape);
 
-    // Get the im2col data buffer
+    // Get the im2col data buffer.
     TBitpacked* packed_im2col_data = reinterpret_cast<TBitpacked*>(im2col_data);
 
     RuntimeShape result_shape;
-    im2col<TBitpacked>(params, packed_input_shape, input_data_bp.data(),
-                       packed_filter_shape, output_shape, im2col_shape,
-                       packed_im2col_data, result_shape, &rhs_data);
+
+    if (read_bitpacked_input) {
+      RuntimeShape packed_input_shape = input_shape;
+
+      const TBitpacked* input_data_bp =
+          reinterpret_cast<const TBitpacked*>(input_data);
+
+      im2col<TBitpacked>(params, packed_input_shape, input_data_bp,
+                         packed_filter_shape, output_shape, im2col_shape,
+                         packed_im2col_data, result_shape, &rhs_data);
+    } else {
+      // Buffer for bitpacked input data.
+      static std::vector<TBitpacked> input_data_bp;
+
+      // The input tensor has this shape which we bitpack along the channels
+      // dimension [batch, input height, input width, channels].
+      RuntimeShape packed_input_shape;
+
+      packbits_tensor(input_shape, input_data, packed_input_shape,
+                      input_data_bp);
+
+      im2col<TBitpacked>(params, packed_input_shape, input_data_bp.data(),
+                         packed_filter_shape, output_shape, im2col_shape,
+                         packed_im2col_data, result_shape, &rhs_data);
+    }
```

Comment:
I don't feel too strongly about this, but what do you think of this:

```suggestion
    RuntimeShape packed_input_shape;
    TBitpacked* input_data_bp;
    if (read_bitpacked_input) {
      packed_input_shape = input_shape;
      input_data_bp =
          reinterpret_cast<const TBitpacked*>(input_data);
    } else {
      // Buffer for bitpacked input data.
      static std::vector<TBitpacked> input_data_bp_buffer;
      // The input tensor has this shape which we bitpack along the channels
      // dimension [batch, input height, input width, channels].
      packbits_tensor(input_shape, input_data, packed_input_shape,
                      input_data_bp_buffer);
      input_data_bp = input_data_bp_buffer.data();
    }
    im2col<TBitpacked>(params, packed_input_shape, input_data_bp,
                       packed_filter_shape, output_shape, im2col_shape,
                       packed_im2col_data, result_shape, &rhs_data);
```

Reply:
This didn't quite work verbatim, but I've done a variant of your suggestion in 51b14d1.

Reply:
I thought this suggestion looked a lot nicer, but it broke things so I've reverted this change :( Can you see anything obviously wrong with 51b14d1?

Reply:
I think in the non-bitpacked block, you don't set `const TBitpacked* input_data_bp` so its not passed to im2col. I'll also give it a try.